### PR TITLE
[WIP] Use Azure pipelines as CI service

### PR DESCRIPTION
**Pull request by liushuyu**
This is a work-in-progress PR since Azure Pipelines not yet has caching mechanisms.

This PR adds CI/CD definitions for use with Azure Pipelines service. 

Known issues: breaks Travis CI configurations.

Currently implemented pipelines:

- Linux Containers:
  - [X] `fresh`
  - [X] ~~frozen~~ cut from current goal
  - [X] `mingw`
  - [X] `clang-format`
  - [ ] Transifex
  - [X] ~~osxcross (PR #4492)~~ cut from current goal
  - [X] Flatpak (PR #4739)

- macOS
  - [X] Xcode build and test

- Windows
  - [X] ~~MSVC~~ cut from current goal

---

### Update on 2019-09-15:

Backported Yuzu's Azure Pipelines configurations to this PR.

Known regression introduced:

- Transifex sync job removed
- MSVC build removed

Known issues:

- Caching system is not working as expected, since [Azure Pipelines' cache is immutable](https://docs.microsoft.com/en-us/azure/devops/pipelines/caching/?view=azure-devops#using-the-cache-task) and [can't be updated or cleared](https://docs.microsoft.com/en-us/azure/devops/pipelines/caching/?view=azure-devops#can-i-clear-a-cache).

---

### Update on 2020-08-27:

Issues resolved:

- MSVC build re-added
- Caching system fixed
- GitHub Release fixed

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4658)

### CODE REVIEW

**Comment by liushuyu** [comment]
Because Citra Emu organization didn't have Azure Pipelines configured, here is the demo:

https://dev.azure.com/liushuyu0110692/liushuyu_011/_build/results?buildId=36

File: azure-pipelines.yml [hunk 1]
```diff
@@ -41,6 +41,29 @@ jobs:
   - bash: /bin/bash -ex ./.travis/${AGENT_JOBNAME}/docker.sh
     displayName: 'Build Citra'
 
+# Transifex translation synchronization job
+- job: LinuxTransifexSync
+  # Azure pipeline's condition expression is very wired
```
Comment (jroweboy) [inline azure-pipelines.yml]: weird*

File: azure-pipelines.yml [hunk 2]
```diff
@@ -41,6 +41,29 @@ jobs:
...
+        eq(variables['Build.Reason'], 'IndividualCI'))
+  pool:
+    vmImage: 'ubuntu-16.04'
+  # we don't use Docker integration here since azure pipeline
+  # Docker integration requires guest Docker image to use GLibc
+  steps:
+  - bash: git submodule update --init --recursive
+    displayName: 'Fetch submodules'
+  - bash: /bin/bash -ex ./.travis/transifex/build.sh
+    # set tx_token variable in the control panel and uncomment the following
```
Comment (jroweboy) [inline azure-pipelines.yml]: so transifex push won't work without adding that var to the control panel?

Comment (liushuyu) [reply azure-pipelines.yml]: Yes. Since you need a access token from Transifex for writing resources of course.

Comment (liushuyu) [reply azure-pipelines.yml]: Also, the only way of setting encrypted variables in Azure Pipelines is to set them in the control panel and click on the lock button next to it

**Comment by liushuyu** [comment]
https://dev.azure.com/liushuyu0110692/liushuyu_011/_build/results?buildId=58

After finishing building, you should see the downloadable artifact under the `Summary` page. (Like this: https://dev.azure.com/liushuyu0110692/liushuyu_011/_build/results?buildId=56&view=results)

The GitHub Release part has to be set up separately through Web UI, and it is quite quirky.

**Comment by jroweboy** [comment]
So moving forward with this, we need to do the following

* Rebase this PR
* Citra-emu org sign up for azure CI pipeline
* Add back in the caching now that azure caching is in preview

Anything else?

File: .ci/scripts/common/post-upload.sh [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+cp license.txt "$REV_NAME"
+cp README.md "$REV_NAME"
+
+tar $COMPRESSION_FLAGS "$ARCHIVE_NAME" "$REV_NAME"
+
+mv "$REV_NAME" $RELEASE_NAME
+
+7z a "$REV_NAME.7z" $RELEASE_NAME
+
+# move the compiled archive into the artifacts directory to be uploaded by travis releases
```
Comment (jroweboy) [inline .ci/scripts/common/post-upload.sh]: by travis releases?

File: .ci/scripts/format/script.sh [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+                 dist/*.svg dist/*.xml; then
+    echo Trailing whitespace found, aborting
+    exit 1
+fi
+
+# Default clang-format points to default 3.5 version one
+CLANG_FORMAT=clang-format-6.0
+$CLANG_FORMAT --version
+
+if [ "$TRAVIS_EVENT_TYPE" = "pull_request" ]; then
```
Comment (jroweboy) [inline .ci/scripts/format/script.sh]: is this an env var on there? (same with TRAVIS COMMIT RANGE below

File: .ci/scripts/macos/docker.sh [hunk 5]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash -ex
+
+brew update
+brew install qt5 sdl2 p7zip ccache ffmpeg ninja
+pip3 install macpack
+
+set -o pipefail
+
+export MACOSX_DEPLOYMENT_TARGET=10.14
```
Comment (jroweboy) [inline .ci/scripts/macos/docker.sh]: Hmm this is still 10.13 right now in travis

**Comment by zhaowenlan1779** [comment]
> Caching system is not working as expected, since Azure Pipelines' cache is immutable and can't be updated or cleared.

I think this can be already done with https://github.com/microsoft/azure-pipelines-tasks/issues/10842#issuecomment-510637798

Basically, you could have a cache key of `branch-name | commit-sha` and ask for **only** `branch-name`  on cache load. This would effective allow a branch-based cache.

Somehow this feature is not documented though

**Comment by liushuyu** [comment]
I have updated this Pull Request again: https://dev.azure.com/liushuyu-011/liushuyu_011/_build/results?buildId=462&view=logs&j=692abbd5-285a-50f1-0358-b6a128ccde90&t=068437d1-623c-5734-6ced-b14effdaaabb

**Comment by liushuyu** [comment]
Superseded by #5602. Closing.

**Pull request closed without merging**